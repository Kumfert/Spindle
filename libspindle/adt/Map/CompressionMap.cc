//
// CompressionMap.cc
//
//  $Id: CompressionMap.cc,v 1.2 2000/02/18 01:31:48 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#include "spindle/CompressionMap.h"

#ifndef SPINDLE_ARCHIVE_H_
#include "spindle/SpindleArchive.h"
#endif

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

#ifdef __FUNC__
#undef __FUNC__
#endif

CompressionMap::CompressionMap() {
  incrementInstanceCount( CompressionMap::MetaData );
  coarse2Fine.take( new MatrixBase );
  reset();
}

bool
CompressionMap::reset() {
  currentState = EMPTY;
  return true;
}

bool 
CompressionMap:: resize( const int nUncompressed, const int nCompressed ) {
  if ( currentState == INVALID ) { return false; }
  if ( nUncompressed <= 0 ) { return false; }
  if ( nCompressed <= 0 ) { return false; }
  if ( nCompressed > nUncompressed ) { return false; }
  szCoarse = nCompressed;
  szFine = nUncompressed;
  fine2Coarse.resize( szFine );
  coarse2Fine->resize( szCoarse, szFine, szFine );
  currentState = UNKNOWN;
  return true;
}

void
CompressionMap::validate() {
  if ( currentState == VALID ) { return; }
  if ( currentState == INVALID ) { return; }
  bool fine2CoarseOkay = false;
  bool coarse2FineOkay = false;
  if ( changedMostRecent == NEITHER ) {
    // pick one
    if ( fine2CoarseOkay = validateFine2Coarse() ) {
      changedMostRecent = FINE2COARSE;
    } else if ( coarse2FineOkay = validateCoarse2Fine() ) {
      changedMostRecent = COARSE2FINE;
    } else {
      currentState = INVALID;
      return;
    }
  }
  if ( changedMostRecent == FINE2COARSE ) {
    // validate fine2coarse if neccessary
    if ( !fine2CoarseOkay ) {
      if ( !validateFine2Coarse() ) {
	currentState = INVALID;
	return;
      }
    }
    // create coarse2fine
    createCoarse2Fine();
  } else if ( changedMostRecent == COARSE2FINE ) {
    // validate coarse2Fine if neccessary
    if ( !coarse2FineOkay ) {
      if ( !validateCoarse2Fine() ) {
	currentState = INVALID;
	return;
      }
    }
    // create fine2coarse
    createFine2Coarse();
  } else {
    // error
    currentState = INVALID;
    return;
  }
  currentState = VALID;
  return;
}

bool
CompressionMap::validateFine2Coarse() {
  // If reported size and actual size differ.
  // the user must have changed the actual size
  // most recent.
  if ( szFine < fine2Coarse.size() ) { 
    currentState = INVALID;
    return false;
  }
  int max=0;
  int stop = fine2Coarse.size();
  const int * cur = fine2Coarse.begin();
  {for ( int i=0; i<stop; ++i, ++cur ) {
    if ( *cur < 0 ) {
      currentState = INVALID;
      return false;
    }
    max = ( *cur > max ) ? *cur : max ;
  }}
  szFine = fine2Coarse.size();
  szCoarse = max + 1; // 0 to n-1
  return true;
}

bool
CompressionMap::validateCoarse2Fine() {
  int nC = coarse2Fine->getColPtr().size() - 1;
  int n = coarse2Fine->getRowIdx().size();
  if ( nC > n ) { 
    currentState = INVALID;
    return false;
  }

  // const int * colPtr = coarse2Fine->getColPtr().lend(); // unused 
  const int * rowIdx = coarse2Fine->getRowIdx().lend();
  SharedArray<int> Temp( n );
  Temp.init( 0 );
  int * temp = Temp.begin();

  // check that each uncompressed vertex is represented.
  {for ( int i=0; i<n; ++i ) {
    temp[ rowIdx[i] ]++;
  }}
  {for ( int i=0; i<n; ++i ) {
    if (temp[ i ] != 1 ) {
      currentState = INVALID;
      return false;
    }
  }}
  szCoarse = nC;
  szFine = n;
  return true;
}


void 
CompressionMap::createCoarse2Fine() {
  int n = szFine;
  int nC = szCoarse;
  const int * fine2coarse = fine2Coarse.begin();

  // first find the max in fine2Coarse;
  coarse2Fine->getColPtr().resize( nC + 1);
  coarse2Fine->getColPtr().init( 0 );
  coarse2Fine->getRowIdx().resize( n );
  SharedArray<int> Temp( n );

  int * colPtr = coarse2Fine->getColPtr().begin();
  int * rowIdx = coarse2Fine->getRowIdx().begin();
  int * temp = Temp.begin();
  
  // now construct reverse map
  {for(int i=0;i<n;++i) { // count instances of each supernode
    colPtr[fine2coarse[i]+1]++;
  }}
  {for(int i=0;i<nC;++i) { // compute a cumulative sum
    colPtr[i+1]+=colPtr[i];
    temp[i] = colPtr[i];  // temp hold the next available spot in head column
  } }
  {for(int i=0;i<n;++i) {
    int j = fine2coarse[i];
    rowIdx[ temp[j]++ ] = i;
  }}
  
  // check
  // {for(i=0;i<nC;++i) {
  //   if( temp[i] != adjhead[i+1]) ERROR(SPINDLE_ERROR, "Indices do not add up!");
  // }}
  return;
}

void 
CompressionMap::createFine2Coarse() {
  int nC = szCoarse;
  //  int n = szFine; // unused 
  fine2Coarse.resize( szFine );
  fine2Coarse.init( -1 );
  const int * colPtr = coarse2Fine->getColPtr().lend();
  const int * rowIdx = coarse2Fine->getRowIdx().lend();
  int * fine2coarse = fine2Coarse.begin();
  {for (int i=0; i<nC; ++i) {
    for (int jj=colPtr[i]; jj<colPtr[i+1]; ++jj ) {
      int j = rowIdx[jj];
      fine2coarse[j]=i;
    }
  }}
}

void
CompressionMap::dump( FILE * fp ) const {
  if ( isValid() ) {
    fprintf(fp,"CompressionMap:");
    const int * p = fine2Coarse.begin(); 
    const int n = fine2Coarse.size();
    for( int i=0; i<n; i++) {
      fprintf(fp,"%d ", p[i] );
    }
  }
}

void
CompressionMap::storeObject( SpindleArchive& ar ) const {
  if ( isValid() ) {
    const int *p = fine2Coarse.lend();
    size_t my_size = fine2Coarse.size();
    pair< const int *, size_t> my_pair( p, my_size );
    ar << my_pair;
  } else { 
    ar << -1;
  }
}

void
CompressionMap::loadObject( SpindleArchive& ar ) {
  size_t new_size = ar.peekLength();
  if ( (int)new_size == -1 ) { 
    int junk;
    ar >> junk;
    currentState = INVALID;
  } else { 
    fine2Coarse.resize( new_size );
    int* p = fine2Coarse.begin();
    pair< int*, size_t > new_pair( p, new_size );
    
    ar >> new_pair;
    
    changedMostRecent = FINE2COARSE;
    validate();
  }
}

SPINDLE_IMPLEMENT_PERSISTANT( CompressionMap, SpindlePersistant )
