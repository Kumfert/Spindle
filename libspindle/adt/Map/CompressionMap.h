//
//  CompressionMap.h -- an abstraction of a compression mapping
//
//  $Id: CompressionMap.h,v 1.2 2000/02/18 01:31:48 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  This class is meant to abstract away the details of a compression
//  mapping.  It has a fine2coarse vector and a coarse2fine matrix.
//  Given one, it will create the other.
//
//  A CompressionMap is not in a valid state until validate() is called
//  to insure that both fine2coarse and coarse2fine are consistent with 
//  each other.
//

#ifndef SPINDLE_COMPRESSION_MAP_H_
#define SPINDLE_COMPRESSION_MAP_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

#ifndef SPINDLE_AUTO_PTR_H_
#include "spindle/SharedPtr.h"
#endif

#ifndef SPINDLE_MATRIX_BASE_H_
#include "spindle/MatrixBase.h"
#endif

#ifndef SPINDLE_PERSISTANT_H_
#include "spindle/SpindlePersistant.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class CompressionMap : public SpindlePersistant {
  SPINDLE_DECLARE_PERSISTANT( CompressionMap )
    
private:
  CompressionMap();
  bool validateFine2Coarse();
  bool validateCoarse2Fine();
  void createFine2Coarse();
  void createCoarse2Fine();
  SharedArray<int>      fine2Coarse;
  SharedPtr<MatrixBase> coarse2Fine;

  enum { NEITHER, FINE2COARSE, COARSE2FINE } changedMostRecent;
  int szCoarse;
  int szFine;
public:
  CompressionMap( const int nUncompressed, const int nCompressed );
  CompressionMap( const CompressionMap& cmap );
  virtual ~CompressionMap();

  int queryNCompressed() const;
  int queryNUncompressed() const;

  const SharedArray<int>& getFine2Coarse() const { return fine2Coarse; }
  SharedArray<int>& getFine2Coarse() { changedMostRecent=FINE2COARSE; return fine2Coarse; }

  const SharedPtr<MatrixBase>& getCoarse2Fine() const { return coarse2Fine; }
  SharedPtr<MatrixBase>& getCoarse2Fine() { changedMostRecent=COARSE2FINE; return coarse2Fine; }

  virtual void validate();
  virtual void dump( FILE * fp ) const;
  virtual bool reset();
  bool resize(  const int nUncompressed, const int nCompressed );

};

inline
int CompressionMap::queryNCompressed() const {
  return szCoarse;
}

inline 
int CompressionMap::queryNUncompressed() const {
  return szFine;
}

inline
CompressionMap::CompressionMap( const int nUncompressed, const int nCompressed ) {
  incrementInstanceCount( CompressionMap::MetaData );
  reset();
  coarse2Fine.take( new MatrixBase );
  resize( nUncompressed, nCompressed );
}

inline
CompressionMap::CompressionMap( const CompressionMap& cmap ) 
  : szCoarse( cmap.szCoarse ), szFine( cmap.szFine ) {
    incrementInstanceCount( CompressionMap::MetaData );
    fine2Coarse.import( cmap.fine2Coarse );
    coarse2Fine.import( cmap.coarse2Fine.operator->() );
    changedMostRecent = cmap.changedMostRecent;
    currentState = cmap.currentState;
}

inline
CompressionMap::~CompressionMap() {
  decrementInstanceCount( CompressionMap::MetaData );
}

SPINDLE_END_NAMESPACE

#endif
