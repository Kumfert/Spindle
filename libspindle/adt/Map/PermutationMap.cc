//
// PermutationMap.cc
//
//  $Id: PermutationMap.cc,v 1.2 2000/02/18 01:31:48 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is povided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  The PermutationMap class is meant to abstract away the concept of
//  a permutation vector which may be of old2New or new2old varieties.
//
//  A PermutationMap is not in a valid state until finalize() is called
//  to insure that both old2New and new2Old are consistent with each
//  other.
//

#include "spindle/PermutationMap.h"

#ifndef SPINDLE_ARCHIVE_H_
#include "spindle/SpindleArchive.h"
#endif

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

#ifdef __FUNC__
#undef __FUNC__
#endif

PermutationMap::PermutationMap() { 
  incrementInstanceCount( PermutationMap::MetaData );
  currentState = EMPTY;
}

void
PermutationMap::validate() {
  if ( currentState == VALID ) { return; }
  int *src; 
  int *dest;
  if ( changedMostRecent == NEITHER ) {
    src = old2New.begin();
    dest = new2Old.begin();
    if ((*src >= 0)&&(*dest <= 0)) { 
      changedMostRecent = OLD2NEW; 
    } else if ((*src <= 0)&&(*dest >= 0)) { 
      changedMostRecent = NEW2OLD; 
      src = new2Old.begin();
      dest = old2New.begin();
    }
  } else if ( changedMostRecent == OLD2NEW ) { 
    src = old2New.begin();
    new2Old.init( 0 );
    dest = new2Old.begin();
  } else { // if ( changedMostRecent == NEW2OLD ) {
    src = new2Old.begin();
    old2New.init( 0 );
    dest = old2New.begin();
  }
  {for (int i=0; i<sz; i++) { 
    int perm_i = src[i];
    if ( perm_i == -1 ) { continue; }
    if ( ( perm_i < 0) ||  ( perm_i >= sz ) ) { 
      currentState = INVALID;
      return; 
    }
    dest[perm_i]++;
  }}
  // ... now make sure each vertex only occurred once
  int * p = dest;
  { for (int i=0; i<sz; i++) { 
    if ( ( *p != 1 ) && (*p != 0) ) { 
      currentState = INVALID;
      return;
    }
  }}

  // ... finally copy the reverse permutation 
  {for (int i=0; i<sz; i++) { 
    dest[ src[ i ] ] = i; 
  }}
  currentState = VALID;
  changedMostRecent = NEITHER;
  return;
}

bool
PermutationMap::reset() {
  old2New.init( -1 );
  new2Old.init( -1 );
  currentState = EMPTY;
  changedMostRecent = NEITHER;
  return true;
}


const SharedArray<int>& 
PermutationMap::getOld2New() const {
  return old2New;
}

SharedArray<int>& 
PermutationMap::getOld2New() {
  if ( (!isValid() ) && (changedMostRecent == NEW2OLD) ) {
    validate();
    assert( isValid() );
  }
  currentState = UNKNOWN;
  changedMostRecent = OLD2NEW;
  return old2New;
}

const SharedArray<int>& 
PermutationMap::getNew2Old() const {
  return new2Old;
}

SharedArray<int>& 
PermutationMap::getNew2Old() {
  if ((!isValid()) && (changedMostRecent == OLD2NEW) ) {
    validate();
    assert( isValid() );
  }
  currentState = UNKNOWN;
  changedMostRecent = NEW2OLD;
  return new2Old;
}

void 
PermutationMap::resize( const int n ) { 
  currentState = EMPTY;
  changedMostRecent = NEITHER;
  sz = n;
  old2New.resize(n);
  new2Old.resize(n);
}

void
PermutationMap::dump( FILE * fp ) const {
  if ( isValid() ) {
    fprintf(fp,"PermutationMap:");
    const int* p = getNew2Old().lend();
    for( int i=0; i<size(); i++) {
      fprintf(fp,"%d ", p[i] );
    }
  }
}

void
PermutationMap::prettyPrint( FILE * fp, bool new2old_perm, bool printLength, 
			     bool onesOffset ) const { 

  if ( !isValid() ) {
    return ;
  }
  int offset = (onesOffset) ? 1 : 0 ;
  
  if ( printLength == true ) {
    fprintf( fp,"%d\n", size() );
  }
  
  const int * p;
  if ( new2old_perm ) { 
    p = new2Old.lend();
  } else { 
    p = old2New.lend();
  }

  for(const int* cur = p, *stop=p + size(); cur <stop; ++cur ) {
    fprintf( fp, "%d\n", *cur + offset);
  }
} 

void
PermutationMap::storeObject( SpindleArchive& ar ) const {
  if ( isValid() ) {
    const int *p = getNew2Old().lend();
    size_t my_size = size();
    pair< const int*, size_t> my_pair( p,my_size );
    ar << my_pair;
  } else {
    ar << -1;
  }
}
#define __FUNC__ "PermutationMap::loadObject( SpindleArchive& ar )"
void
PermutationMap::loadObject( SpindleArchive& ar ) {
  size_t new_size = ar.peekLength();
  if ( (int)new_size == -1 ) { 
    int junk;
    ar >> junk;
    currentState = INVALID;
  } else {
    resize(new_size);
    int* p = getNew2Old().begin();
    pair<int*, size_t> new_pair( p, new_size );
    
    ar >> new_pair;
    
    WARN_IF( new_pair.second != new_size, "Warning, expected size=%d, but extracted only %d.", 
	     new_pair.second, new_size );
    changedMostRecent = NEW2OLD;
    validate();
  }
}
#undef __FUNC__

SPINDLE_IMPLEMENT_PERSISTANT( PermutationMap, SpindlePersistant )


