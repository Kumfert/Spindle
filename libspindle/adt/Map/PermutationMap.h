//
//  PermutationMap.h -- an abstraction of a permutation vector.
//
//  $Id: PermutationMap.h,v 1.2 2000/02/18 01:31:49 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  The PermutationMap class is meant to abstract away the concept of
//  a permutation vector which may be of old2New or new2old varieties.
//
//  A PermutationMap is not in a valid state until validate() is called
//  to insure that both old2New and new2Old are consistent with each
//  other.
//

#ifndef SPINDLE_PERMUTATION_H_
#define SPINDLE_PERMUTATION_H_

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifndef SPINDLE_AUTO_ARRAY_H_
#include "spindle/SharedArray.h"
#endif

#ifndef SPINDLE_PERSISTANT_H_
#include "spindle/SpindlePersistant.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class PermutationMap : public SpindlePersistant {
  SPINDLE_DECLARE_PERSISTANT( PermutationMap )

private:
  PermutationMap();
  SharedArray<int> old2New;
  SharedArray<int> new2Old;

  enum { NEITHER, OLD2NEW, NEW2OLD } changedMostRecent;
  int sz;
public:
  // constructors/destructors
  PermutationMap( const int n );
  PermutationMap( const PermutationMap& perm );
  virtual ~PermutationMap();

  int size() const;

  // access specific permuation vectors
  const SharedArray<int>& getOld2New() const;
  SharedArray<int>& getOld2New();

  const SharedArray<int>& getNew2Old() const;
  SharedArray<int>& getNew2Old();


  virtual void validate();
  virtual void dump( FILE * stream ) const;

  virtual bool reset();
  void resize( const int n );
  void prettyPrint( FILE * fp, bool new2old_perm=true, bool printLength=false, 
		    bool onesOffset=false ) const; 

};

inline 
int PermutationMap::size() const {
  return sz;
}

inline 
PermutationMap::PermutationMap( const int n )
  : old2New(n), new2Old(n), sz(n) {
    incrementInstanceCount( PermutationMap::MetaData );
    reset();
}

inline 
PermutationMap::PermutationMap( const PermutationMap& perm ) 
  : sz( perm.size() ) {
    incrementInstanceCount( PermutationMap::MetaData );
    old2New.import( perm.old2New );
    new2Old.import( perm.new2Old );
    changedMostRecent = perm.changedMostRecent;
    currentState = perm.currentState;
}

inline 
PermutationMap::~PermutationMap() {
  decrementInstanceCount( PermutationMap::MetaData );
}

SPINDLE_END_NAMESPACE

#endif


