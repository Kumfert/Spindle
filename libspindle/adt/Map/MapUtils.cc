//
// MapUtils.h
//
//  $Id: MapUtils.cc,v 1.2 2000/02/18 01:31:48 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#include "spindle/MapUtils.h"

#ifndef SPINDLE_PERMUTATION_H_
#include "spindle/PermutationMap.h"
#endif

#ifndef SPINDLE_COMPRESSION_MAP_H_
#include "spindle/CompressionMap.h"
#endif 

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

PermutationMap*
MapUtils::createUncompressedPermutationMap( const PermutationMap* perm, const CompressionMap* cmap ) {

  if ( (!perm->isValid()) || (!cmap->isValid()) ) { return 0; }

  const int n = cmap->queryNUncompressed();
  const int nC = cmap->queryNCompressed();
  
  if ( perm->size() != nC ) { return 0; }
  
  PermutationMap* newPerm = new PermutationMap(n); 
  int* old2new = newPerm->getOld2New().begin();
 
  const int * new2old = perm->getNew2Old().lend();
  const int * colPtr = cmap->getCoarse2Fine()->getColPtr().lend();
  const int * rowIdx = cmap->getCoarse2Fine()->getRowIdx().lend();
    
  // the ordering for the vertices in each supernode
  int nextidx = 0; 
  {for(int ii=0;ii<nC;++ii) {
    int i = new2old[ ii ];
    for(int jj=colPtr[i]; jj<colPtr[i+1]; ++jj ) {
      int j = rowIdx[jj];
      old2new[j] = nextidx++;
    }
  }}

  newPerm->validate();
  if ( newPerm->isValid() ) {
    return newPerm;
  } else {
    delete newPerm;
    return 0;
  }
}

