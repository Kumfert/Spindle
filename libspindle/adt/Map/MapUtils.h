//
// MapUtils.h
//
//  $Id: MapUtils.h,v 1.2 2000/02/18 01:31:48 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#ifndef SPINDLE_MAP_UTILS_H_
#define SPINDLE_MAP_UTILS_H_

#ifndef SPINDLE_PERMUTATION_MAP_H_
#include "spindle/PermutationMap.h"
#endif

#ifndef SPINDLE_COMPRESSION_MAP_H_
#include "spindle/CompressionMap.h"
#endif

SPINDLE_BEGIN_NAMESPACE

class MapUtils { 
public:
  static PermutationMap * createUncompressedPermutationMap( const PermutationMap* perm, const CompressionMap* cmap ) ;
};

SPINDLE_END_NAMESPACE

#endif
