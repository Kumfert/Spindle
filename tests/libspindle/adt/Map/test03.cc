//
// src/adt/Map/drivers/tests/test03.cc -- test MapUtils
//
//  $Id: test03.cc,v 1.2 2000/02/18 01:32:07 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify, distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is povided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//


#include "spindle/CompressionMap.h"
#include "spindle/PermutationMap.h"
#include "spindle/MapUtils.h"

#ifndef SPINDLE_SYSTEM_H_
#include "spindle/SpindleSystem.h"
#endif

#include <iostream.h>
#include <iomanip.h>

#define REGISTRY SpindleSystem::registry()

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

int main() {
  ios::sync_with_stdio();
  SpindleSystem::registry().setDumpOnDestroy( stdout );
  SpindleSystem::setErrorFile( stdout );

  int comp_array[] = { 0, 0, 1, 0, 1, 1, 2, 3, 4, 4, 3, 4 };
  CompressionMap cmap(12,5);
  cmap.getFine2Coarse().import( comp_array, 12 );
  cmap.validate();
  if ( ! cmap.isValid() ) { 
    cerr << "WARNING: cannot create valid CompressionMap." << endl;
  }

  int perm_array[] = { 2, 3, 0, 1, 4 }; // old2new ordering by minimum supernode weight
  PermutationMap perm(5);
  perm.getNew2Old().import( perm_array, 5);
  perm.validate();
  if ( ! perm.isValid() ) {
    cerr << "WARNING: cannot create valid PermutationMap." << endl;
  }

  PermutationMap* newPerm = MapUtils::createUncompressedPermutationMap( &perm, &cmap );
  if ( (newPerm == 0) || ( !newPerm->isValid() ) ) { 
    cerr << "ERROR: could not create an uncompressed permutation" << endl;
  } else {

    cout << "This new permutation object is serial #" << newPerm->queryIDNumber() << endl;
    
    cout << "Dumping the uncompressed PermutationMap { " << endl << "  ";
    newPerm->dump( stdout );
    cout << endl << "}" << endl;
    
    delete newPerm;
  }

}
