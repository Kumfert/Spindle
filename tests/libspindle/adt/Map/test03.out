This new permutation object is serial #2
Dumping the uncompressed PermutationMap { 
  PermutationMap:6 7 10 0 1 3 2 4 5 8 9 11 
}
Destroying Registry... PersistanceRegistery::dump():  4 items
             CompressionMap (44) 1/1---> SpindlePersistant
             PermutationMap (44) 2/2---> SpindlePersistant
           SpindleBaseClass (8) 3/3---> 
          SpindlePersistant (12) 3/3---> SpindleBaseClass
